File: Course-Registration-System-main/java-(iteration_1)/src/TranscriptSaver.java
import java.io.FileWriter;
import java.io.IOException;

public class TranscriptSaver {

    // Method to save a transcript object to a JSON file
    public void save(Transcript transcript) {
        // Define the file path for the transcript based on its ID
        String filePath = "Data/Transcripts/transcript" + transcript.getId() + ".json";
        try (FileWriter fileWriter = new FileWriter(filePath)) {
            StringBuilder jsonTranscript = new StringBuilder();
            
            // Build the JSON structure for the transcript
            jsonTranscript.append("{\n")
                    .append("  \"id\": ").append(transcript.getId()).append(",\n")
                    .append("  \"GPA\": ").append(transcript.getGPA()).append(",\n");

            // Check if there are grades in the transcript
            if (transcript.getGrades() != null && !transcript.getGrades().isEmpty()) {
                jsonTranscript.append("  \"grades\": [\n");
                
                // Iterate through each grade and add it to the JSON structure
                for (int i = 0; i < transcript.getGrades().size(); i++) {
                    Grade grade = transcript.getGrades().get(i);
                    jsonTranscript.append("    {\n")
                            .append("      \"courseName\": \"").append(grade.getCourseName()).append("\",\n")
                            .append("      \"courseCode\": \"").append(grade.getCourseCode()).append("\",\n")
                            .append("      \"semester\": \"").append(grade.getSemester()).append("\",\n")
                            .append("      \"credit\": ").append(grade.getCredit()).append(",\n")
                            .append("      \"numericGrade\": ").append(grade.getNumericGrade()).append(",\n")
                            .append("      \"letterGrade\": \"").append(grade.getLetterGrade()).append("\"\n")
                            .append("    }");

                    // Add a comma if there are more grades to be added, otherwise end the JSON array
                    if (i < transcript.getGrades().size() - 1) {
                        jsonTranscript.append(",\n");
                    } else {
                        jsonTranscript.append("\n");
                    }
                }
                jsonTranscript.append("  ]\n");
            } else {
                // If there are no grades, add an empty grades array to the JSON structure
                jsonTranscript.append("  \"grades\": []\n");
            }

            jsonTranscript.append("}");

            // Write the constructed JSON transcript to the file
            fileWriter.write(jsonTranscript.toString());
        } catch (IOException e) {
            // If an exception occurs during file writing, display an error message
            System.err.println("Failed to save transcript to file: " + e.getMessage());
            // logger.error("Failed to save transcript to file", e); - This might be a log statement for logging the error.
        }
    }
}

File: Course-Registration-System-main/python-(iteration_2)/src/TranscriptSaver.py
import json
import logging

class TranscriptSaver:
    def __init__(self):
        self.__file_path = ""
        self.logger = logging.getLogger(__name__)
        logging.basicConfig(level=logging.INFO)
        
    def save(self, transcript):
        # Define the file path for the transcript based on its ID
        self.__file_path = f"../Data/Transcripts/transcript{transcript.getId()}.json"
        try:
            with open(self.__file_path, 'w') as file_writer:
                json_transcript = {
                    "id": transcript.getId(),
                    "GPA": transcript.getGPA(),
                    "grades": []
                }
                self.logger.info("Transcript is changed for student with ID: %d", transcript.getId())

                # Check if there are grades in the transcript
                if transcript.getGrades() and len(transcript.getGrades()) > 0:
                    for grade in transcript.getGrades():
                        grade_data = {
                            "courseName": grade.getCourseName(),
                            "courseCode": grade.getCourseCode(),
                            "semester": grade.getSemester(),
                            "credit": grade.getCredit(),
                            "numericGrade": grade.getNumericGrade(),
                            "letterGrade": grade.getLetterGrade()
                        }
                        json_transcript["grades"].append(grade_data)
                
                # Write the constructed JSON transcript to the file
                file_writer.write(json.dumps(json_transcript, indent=2))
        except IOError as e:
            # If an exception occurs during file writing, display an error message
            print(f"Failed to save transcript to file: {e}")

    def getFilePath(self):
        return self.__file_path
